# Step 4: Segregating E gene and Whole Genomes and performing quaility control
configfile: "config/config.yaml"
indir = config["input_namespace"]
outdir = config["output_namespace"]

serotype = ['Dengue_1', 'Dengue_2', 'Dengue_3', 'Dengue_4']

rule split_genome_and_QC:
    input:
        fasta_files = expand("results/{indir}/Aligned_{serotype}/nextalign.aligned.fasta", indir=indir, serotype=serotype),
        script = srcdir("../resources/scripts/Seperate_EG_and_WG.R"),
    output:
        E_gene_dir = expand("results/{outdir}/{serotype}_EG.fasta", outdir=outdir, serotype=serotype),
        WG_gene_dir = expand("results/{outdir}/{serotype}_WG.fasta", outdir=outdir, serotype=serotype),
    params:
        wg_threshold = 0.29,
        eg_threshold = 0.05,
        indir = expand("results/{indir}", indir=indir),
        outdir = expand("results/{outdir}", outdir=outdir),
    log:
        "logs/process_dengue_data.log"
    message:
        "Segregating E gene and whole genomes from aligned Dengue virus sequences and performing quality control."
    shell:
        """
        Rscript {input.script} --WG_threshold {params.wg_threshold} --EG_threshold {params.eg_threshold} --input_dir {params.indir} --output_dir {params.outdir}
        """
